pydantic_ai_slim/pydantic_ai/_agent_graph.py

```diff
@@ -509,6 +510,10 @@ async def _prepare_request(
         # Update the new message index to ensure `result.new_messages()` returns the correct messages
         ctx.deps.new_message_index -= len(original_history) - len(message_history)
 
+        prompt_templates = ctx.deps.prompt_templates
+        if prompt_templates:
+            _apply_prompt_templates(message_history, prompt_templates, run_context)
```

Comment:
We should not modify the `ModelRequest`s or parts in the history in place, so please have this method build new object and a new list.

We should make sure the modified final `ModelRequest` shows up in `result.all_messages()` etc, so we need `ctx.state.message_history[:] = message_history` as above